### PR TITLE
Replace more implicit array to bool conversions

### PR DESCRIPTION
**Pull request by yebblies**
Needed for https://github.com/D-Programming-Language/dmd/pull/2885

### CODE REVIEW

File: std/string.d [hunk 1]
```diff
@@ -3175,7 +3175,7 @@ body
         assert(translate(to!S("hello world"), makeTrans("hl", "q5")) == to!S("qe55o wor5d"));
         assert(translate(to!S("hello \U00010143 world"), makeTrans("hl", "q5")) ==
                to!S("qe55o \U00010143 wor5d"));
-        assert(translate(to!S("hello world"), makeTrans("ol", "1o")), to!S("heool wlrdd"));
+        assert(translate(to!S("hello world"), makeTrans("ol", "1o")) == to!S("heoo1 w1rod"));
```
Comment (yebblies) [inline std/string.d]: As far as I can see this was a bug in the unittests, the result string's content wasn't checked.

**Comment by JakobOvrum** [comment]
LGTM

**Comment by JakobOvrum** [comment]
Auto-merge toggled on

**Comment by andralex** [comment]
I continue being unclear and mildly unhappy about this. So now we need to write `if (a == null)` instead of `if (a)`. Why? Now we have inconsistency with pointers and references where comparison against null was impled.

@yebblies could you please explain, thanks

**Comment by quickfur** [comment]
Actually, wasn't it sometime last year that pointers can no longer be used in boolean expressions unless you explicitly tested them with `!is null`? I remember being mildly annoyed when my code broke because of that, but afterwards I actually appreciated it because it made my code more readable -- the intent became overt instead of merely implied.

**Comment by yebblies** [comment]
> I continue being unclear and mildly unhappy about this. So now we need to write if (a == null) instead of if (a). Why? Now we have inconsistency with pointers and references where comparison against null was impled.
>
> @yebblies could you please explain, thanks

Because currently, `if (arr)` only checks that `arr.ptr` is non-null, instead of considering the length.  This means an empty slice of an array will evaluate as true.  See my comment in https://github.com/D-Programming-Language/dmd/pull/2885#issuecomment-71964785 and https://d.puremagic.com/issues/show_bug.cgi?id=4733 for more.

**Comment by schveiguy** [comment]
> So now we need to write if (a == null) instead of if (a).

Note than in every instance here, if(a == null) is NOT the same operation as if(a). So in every case we are fixing a potential bug.

take for instance the std.json one:

```D
return str.data ? str.data : "";
```

str is an appender for a string. Note that if str.data has been pre-allocated, but hasn't actually been appended to, then this function will *still* return str.data, which is likely a pointer into some heap data that doesn't need to be kept around for returning a zero-length slice of it. Returning instead the static "" will let the GC collect it. Checking str.data.length instead is the right call.

I think there is something misleading with changing the statement to if(a == null) note that this is equivalent (essentially) to if(a.length == 0). Using the word "null" seems to imply you are checking the pointer, which should never happen.

**Pull request merged**